File: src/flags.rs
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum NIDenType {
    RHO,
    SIGMA,
    TAU,
    LAPL,
}

impl NIDenType {
    /// Returns the number of components in the output density for this XC type.
    ///
    /// - RHO: 1 component (density)
    /// - SIGMA: 4 components (density + 3 gradient components)
    /// - TAU: 5 components (density + 3 gradient components + kinetic energy density)
    /// - LAPL: 6 components (density + 3 gradient components + kinetic energy density + laplacian)
    pub fn num_rho_components(&self) -> usize {
        match self {
            NIDenType::RHO => 1,
            NIDenType::SIGMA => 4,
            NIDenType::TAU => 5,
            NIDenType::LAPL => 6,
        }
    }

    /// Returns the required AO derivative level for this XC type.
    ///
    /// - RHO: 0th order
    /// - SIGMA: 1st order (gradient)
    /// - TAU: 1st order (gradient)
    /// - LAPL: 2nd order (Laplacian)
    pub fn num_ao_deriv(&self) -> usize {
        match self {
            NIDenType::RHO => 0,
            NIDenType::SIGMA => 1,
            NIDenType::TAU => 1,
            NIDenType::LAPL => 2,
        }
    }
}

File: pynimatmul/flags.py
def num_nvar(den_type: str) -> int:
    """Number of density variables for a given density type.

    Parameters
    ----------
    den_type : str
        The density type. One of "RHO", "SIGMA", "TAU", "LAPL".

    Returns
    -------
    int
        The number of density variables.
    """
    mapping = {
        "RHO": 1,
        "SIGMA": 4,
        "TAU": 5,
        "LAPL": 6,
    }
    if den_type not in mapping:
        raise ValueError(f"Unsupported density type: {den_type}")
    return mapping[den_type]


def num_ao_comp(den_type: str) -> int:
    """Number of AO components for a given density type.

    Parameters
    ----------
    den_type : str
        The density type. One of "RHO", "SIGMA", "TAU", "LAPL".

    Returns
    -------
    int
        The number of AO components.
    """
    mapping = {
        "RHO": 1,
        "SIGMA": 4,
        "TAU": 4,
        "LAPL": 10,
    }
    if den_type not in mapping:
        raise ValueError(f"Unsupported density type: {den_type}")
    return mapping[den_type]


def num_ao_deriv(den_type: str) -> int:
    """AO derivative order for a given density type.

    Parameters
    ----------
    den_type : str
        The density type. One of "RHO", "SIGMA", "TAU", "LAPL".

    Returns
    -------
    int
        The AO derivative order.
    """
    mapping = {
        "RHO": 0,
        "SIGMA": 1,
        "TAU": 1,
        "LAPL": 2,
    }
    if den_type not in mapping:
        raise ValueError(f"Unsupported density type: {den_type}")
    return mapping[den_type]
